src/LineSegment.cpp

```diff
@@ -2,154 +2,209 @@
 // Created by rolf on 15-5-19.
 //
 
-#include "../include/roboteam_utils/LineSegment.h"
 #include "../include/roboteam_utils/Line.h"
+#include "../include/roboteam_utils/LineSegment.h"
 
 namespace rtt {
-    bool LineSegment::doesIntersect(const Line &line) const {
-        if (intersects(line)) {
-            return true;
-        }
-        return false;
+LineSegment::LineSegment(const Line &line) {
+    start=line.start;
+    end=line.end;
+}
+double LineSegment::length() const {
+    return (end - start).length();
+}
+
+double LineSegment::length2() const {
+    return (end - start).length2();
+}
+
+double LineSegment::slope() const {
+    return (end.y - start.y)/(end.x - start.x);
+}
+
+bool LineSegment::isVertical() const {
+    return (end.x == start.x) && (end.y != start.y);
+}
+
+Vector2 LineSegment::direction() const {
+    return Vector2(end - start);
+}
+
+double LineSegment::intercept() const {
+    return start.y - this->slope()*start.x;
+}
+
+std::pair<double, double> LineSegment::coefficients() const {
+    return {slope(), intercept()};
+}
+
+bool LineSegment::isPoint() const {
+    return start == end;
+}
+
+bool LineSegment::isParallel(const LineSegment &line) const {
+    // check if line is vertical, otherwise check the slope
+    if (this->isVertical() || line.isVertical()) {
+        return this->isVertical() && line.isVertical();
     }
+    return this->slope() == line.slope();
+}
+bool LineSegment::isParallel(const Line &line) const {
+    // check if line is vertical, otherwise check the slope
+    if (this->isVertical() || line.isVertical()) {
+        return this->isVertical() && line.isVertical();
+    }
+    return this->slope() == line.slope();
+}
+bool LineSegment::doesIntersect(const Line &line) const {
+    if (intersects(line)) {
+        return true;
+    }
+    return false;
+}
 
 //this is the algorithm from
 // https://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect
 //takes overlaps into account in contrast to the intersect() function
-    bool LineSegment::doesIntersect(const LineSegment &line) const {
-        Vector2 p = start, q = line.start, r = direction(), s = line.direction();
-        double denom = r.cross(s);
-        double numer = (q - p).cross(r);
-        if (denom == 0) {
-            if (numer == 0) {
-                // lines are colinear
-                double t0 = (q - p).dot(r) / r.length2();
-                double t1 = t0 + s.dot(r) / r.length2();
-                if (t0 < 0) {
-                    return t1 >= 0;
-                } else if (t0 > 1) {
-                    return t1 <= 1;
-                }
-                return true;
+bool LineSegment::doesIntersect(const LineSegment &line) const {
+    Vector2 p = start, q = line.start, r = direction(), s = line.direction();
+    double denom = r.cross(s);
+    double numer = (q - p).cross(r);
+    if (denom == 0) {
```

Comment:
too many indents tbh, see if you can refactor this function a bit, it's quite long tbh

Reply:
There isn't much to do about it. It's a complicated function but I am personally fine with that as long as they are contained to math problems like these, as we also want these functions to be performance friendly in general.